File: Data Structures & Algorithms/house-robber-ii/submission-2.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.size() == 1) {
            return nums[0];
        }
        int n = nums.size();
        // left
        int left_max = 0;
        vector<int> dp(n, 0);
        dp[0] = nums[0];
        dp[1] = max(nums[0], nums[1]);
        for(int i = 2; i < n - 1;  ++i) {
            dp[i] = max(nums[i] + dp[i - 2], dp[i - 1]);
        }
        left_max = dp[n - 2];

        // right
        int right_max = 0;
        dp[n - 1] = nums[n - 1];
        dp[n - 2] = max(nums[n - 1], nums[n - 2]);
        for(int i = n - 3; i > 0; --i) {
            dp[i] = max(nums[i] + dp[i + 2], dp[i + 1]);
        }
        right_max = dp[1];

        return max(left_max, right_max);
    }
};

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-2.py
from collections import deque

class Solution:
    def largestRectangleArea(self, heights: list[int]) -> int:
        n = len(heights)
        stack = deque()
        max_area = 0

        left, right = [0] * n, [0] * n

        for i in range(n):
            while stack and heights[stack[-1]] >= heights[i]:
                stack.pop()
            left[i] = stack[-1] if stack else -1
            stack.append(i)

        stack.clear()

        for i in range(n - 1, -1, -1):
            while stack and heights[stack[-1]] >= heights[i]:
                stack.pop()
            right[i] = stack[-1] if stack else n
            stack.append(i)
        
        for i in range(n):
            width = right[i] - left[i] - 1
            max_area = max(max_area, heights[i] * width)

        return max_area